### PR TITLE
Add PyTorch backend for MF

### PR DESCRIPTION
**Pull request by hieuddo**
### Description

  * add PyTorch backend for MF
  * add TorchMF to `examples/biased_mf.py`
  * MF minor fix: it should be either `user_idx` or `item_idx` is unknown?
https://github.com/PreferredAI/cornac/blob/b44e19c9f6a31991698c9306f19518e593a51ecc/cornac/models/mf/recom_mf.pyx#L269

### Checklist:

- [ ] I have added tests.

### CODE REVIEW

**Comment by hieuddo** [comment]
Results obtained from `examples/biased_mf.py`:
|          |    MAE |   RMSE | NDCG@10 | Recall@10 | Train (s) | Test (s) |
|----------|--------|--------|---------|-----------|-----------|----------|
| GlobalAvg| 0.9408 | 1.0835 |  0.0085 |    0.0023 |    0.0000 |   8.1069 |
| MF       | 0.6910 | 0.8470 |  0.1058 |    0.0335 |    0.4068 |  36.5108 |
| TorchMF  | 0.6947 | 0.8590 |  0.1014 |    0.0308 |  119.5144 |  10.5325 |

|          |    MAE |   RMSE | NDCG@10 | Recall@10 | Train (s) | Test (s) |
|----------|--------|--------|---------|-----------|-----------|----------|
| GlobalAvg| 0.9373 | 1.0797 |  0.0088 |    0.0026 |    0.0000 |   9.0318 |
| MF       | 0.6911 | 0.8473 |  0.0976 |    0.0318 |    0.4288 |  32.9461 |
| TorchMF  | 0.6962 | 0.8608 |  0.0977 |    0.0313 |  120.6941 |  11.7570 |

File: cornac/models/mf/torchmf.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        if "V" in init_params:
+            self.i_factors.weight.data = torch.from_numpy(init_params["V"])
+        if "Bu" in init_params:
+            self.u_biases.weight.data = torch.from_numpy(init_params["Bu"])
+        if "Bi" in init_params:
+            self.i_biases.weight.data = torch.from_numpy(init_params["Bi"])
+
+    def forward(self, uids, iids):
+        ues = self.u_factors(uids)
+        uis = self.i_factors(iids)
```
Comment (lthoang) [inline cornac/models/mf/torchmf.py:62]: do you mean `ies` (item embeddings) here?

**Comment by tqtg** [comment]
@hieuddo I did some code refactoring to merge them into one MF model with different options of backends. The results appear to be similar to yours.
```
TEST:
...
           |    MAE |   RMSE | NDCG@10 | Recall@10 | Train (s) | Test (s)
---------- + ------ + ------ + ------- + --------- + --------- + --------
GlobalAvg  | 0.9426 | 1.0868 |  0.0093 |    0.0036 |    0.0000 |   6.3037
MF-cpu     | 0.6914 | 0.8494 |  0.0899 |    0.0293 |    0.1364 |   9.3481
MF-pytorch | 0.7023 | 0.8579 |  0.1422 |    0.0440 |   68.6235 |   9.4482
```

**Comment by hieuddo** [comment]
Looks great.

One minor, though: should we unify formatter settings to reduce changes due to different formatters?

I'm currently using `black`, `autopep8`, and `isort` altogether, with `black` as the default formatter. I think `black` is the priority because very few configurations are allowed, which makes code look the same every time; while `autopep8` ignores some parts and only formats where the code doesn't follow `pep8` style.

**Comment by tqtg** [comment]
@hieuddo I'm also using `black`, and we do check PEP8 style(see [Link with flake8](https://github.com/PreferredAI/cornac/actions/runs/6951969034/job/18914800828?pr=546) step). While trying to enforce convention for our core codebase, we're a bit more flexible with models to ease the contribution.

**Comment by hieuddo** [comment]
[like]  DO Dinh Hieu reacted to your message:
________________________________

**Review by tqtg — APPROVED** [review]

**Pull request merged**